File: src/main/java/de/twometer/orion/render/Camera.java
package de.twometer.orion.render;

import de.twometer.orion.core.OrionApp;
import de.twometer.orion.gl.Window;
import de.twometer.orion.util.MathF;
import org.joml.Matrix4f;
import org.joml.Vector2f;
import org.joml.Vector3f;

public class Camera {

    private final Vector3f position = new Vector3f(0, 0, 0);

    private final Vector2f angle = new Vector2f(0, 0);

    private Matrix4f viewMatrix;

    private Matrix4f projectionMatrix;

    private float fov = MathF.toRadians(70);

    private float near = 0.1f;

    private float far = 100.0f;

    public void update() {
        Window window = OrionApp.get().getWindow();
        float aspect = window.getWidth() / (float) window.getHeight();

        float yaw = MathF.toRadians(angle.x);
        float pitch = MathF.toRadians(angle.y);

        Vector3f direction = new Vector3f(
                MathF.cos(pitch) * MathF.sin(yaw),
                MathF.sin(pitch),
                MathF.cos(pitch) * MathF.cos(yaw)
        );
        Vector3f right = new Vector3f(
                MathF.sin(yaw - MathF.PI / 2.0f),
                0,
                MathF.cos(yaw - MathF.PI / 2.0f)
        );
        Vector3f up = new Vector3f(right).cross(direction);

        viewMatrix = new Matrix4f().lookAt(position, direction.add(position), up);
        projectionMatrix = new Matrix4f().perspective(fov, aspect, near, far);
    }

    public Matrix4f getViewMatrix() {
        return viewMatrix;
    }

    public Matrix4f getProjectionMatrix() {
        return projectionMatrix;
    }

    public Vector3f getPosition() {
        return position;
    }

    public Vector2f getAngle() {
        return angle;
    }

    public void setFov(float fov) {
        this.fov = MathF.toRadians(fov);
    }

    public void setNear(float near) {
        this.near = near;
    }

    public void setFar(float far) {
        this.far = far;
    }
}

File: neko-engine/src/main/kotlin/de/twometer/neko/scene/Camera.kt
package de.twometer.neko.scene

import de.twometer.neko.core.NekoApp
import de.twometer.neko.util.MathF
import de.twometer.neko.util.MathF.cos
import de.twometer.neko.util.MathF.sin
import org.joml.Matrix4f
import org.joml.Vector2f
import org.joml.Vector3f
import org.joml.Vector4f


class Camera {

    val position = Vector3f(1f, 1f, 1f)
    val rotation = Vector2f()

    val viewMatrix = Matrix4f()
    val projectionMatrix = Matrix4f()
    private val projViewMatrix = Matrix4f()
    private val frustumPlanes = Array(6) { Vector4f() }

    var fov = MathF.toRadians(70.0f)
    var zNear = 0.1f
    var zFar = 150.0f

    lateinit var direction: Vector3f
        private set
    lateinit var right: Vector3f
        private set
    lateinit var up: Vector3f
        private set

    fun update() {
        val (width, height) = NekoApp.the.window.getSize()
        val aspect = width.toFloat() / height.toFloat()

        val yaw = rotation.x
        val pitch = rotation.y

        direction = Vector3f(
            cos(pitch) * sin(yaw),
            sin(pitch),
            cos(pitch) * cos(yaw)
        )

        right = Vector3f(
            sin(yaw - MathF.PI / 2.0f),
            0.0f,
            cos(yaw - MathF.PI / 2.0f)
        )

        up = Vector3f(right).cross(direction)

        viewMatrix.identity().lookAt(position, Vector3f(position).add(direction), up)
        projectionMatrix.identity().perspective(fov, aspect, zNear, zFar)

        projViewMatrix.set(projectionMatrix).mul(viewMatrix)
        for (i in 0..5)
            projViewMatrix.frustumPlane(i, frustumPlanes[i])
    }

    fun isInFrustum(vec: Vector3f, radius: Float): Boolean {
        return frustumPlanes.none { it.x * vec.x + it.y * vec.y + it.z * vec.z + it.w <= -radius }
    }

}